src/planning/include/planning/cone_coloring.hpp

```diff
@@ -0,0 +1,200 @@
+
+  /**
+    * @brief squared distance between two cones
+    * 
+    * @param c1 first cone
+    * @param c2 second cone
+    * @return double squared distance between the two cones
+    */
+  double squaredDistance(Cone *c1, Cone *c2);
+
+  /**
+   * @brief place the next cone in the path
+   * 
+   * @param un_visited_cones array of possible cones to be visited (marked with false if not visited yet)
+   * @param distance_threshold maximum distance to reach the next cone
+   * @param side place the next cone in the left [true] or right [false] side
+   * @return true if the cone has been successfully placed
+   * @return false if the cone hasn't been placed
+   */
+  bool placeNextCone(std::vector<std::pair<Cone *, bool> *> &visited_cones, double distance_threshold, int n, bool side);
+  
+  /**
+   * @brief color the cones based on the cost to reach them
+   * 
+   * @param input_cones cones to be colored
+   * @param initial_car_pose initial pose of the car
+   * @param distance_threshold maximum allowed distance to reach the next cone
+   */
+  void colorCones(std::vector<Cone *> input_cones, Pose initial_car_pose, double distance_threshold);
+
+  /**
+   * @brief get an estimate to the distance to the left side of the track
+   * 
+   * @param initial_car_pose 
+   * @param side distance to the left [true] or right [false] side
+   * @param x coordinate of the point to estimate the distance
+   * @param y coordinate of the point to estimate the distance
+   * @return double estimate of the distance to the left side track
+   * 
+   * @note assumming the car is at the initial_car_pose and going straight, the right side of the track can be estimated to be 
+   * defined by:  y - (car_y + 2*cos(θ)) - tan(θ)*(x - (car_x  + 2*sen(θ))) = 0
+   * and the right side by : y - (car_y - 2*cos(θ)) - tan(θ)*(x - (car_x  - 2*sen(θ))) = 0
+   * This also assumes the track width to be 4 meters since the minimum is 3 and the maximum is 5
+   */
+  double distance_to_side(Pose initial_car_pose, bool side, double x, double y);
+
+  /**
+   * @brief get the initial cone to start the path from side determined by parameter
+   * 
+   * @param candidates possible cones to start the path (cones closer than threshold)
+   * @param initial_car_pose initial pose of the car
+   * @param side reach the cone form the left [true] or right [false] side
+   * @return Cone* initial cone to start the path
+   */
+  Cone* getInitialCone(std::vector<Cone *> candidates, Pose initial_car_pose, bool side);
+
+  /**
+   * @brief make a vector of cones and a boolean to determine if the cone has been visited
+   * 
+   * @param cones cones to be marked as visited [true] or unvisited [false]
+   * @return std::vector<std::pair<Cone *, bool> *> vector of cones and a boolean to determine if the cone has been visited
+   */ 
+  std::vector<std::pair<Cone *, bool> *> makeUnvisitedCones(std::vector<Cone *> cones);
```

Comment:
Try to use smart pointers (std::shared_ptr) instead of normal pointers for cones